internal/gke/purge.go

```diff
@@ -0,0 +1,791 @@
+	client, err := privateca.NewCertificateAuthorityClient(ctx, config.ClientOptions()...)
+	if err != nil {
+		return nil, doErr(err)
+	}
+	op, err := client.DeleteCertificateAuthority(ctx, &privatecapb.DeleteCertificateAuthorityRequest{
+		Name: path,
+	})
+	if err != nil {
+		client.Close()
+		return nil, doErr(err)
+	}
+	fmt.Fprintf(os.Stderr, "Done\n")
+	return func() error {
+		defer client.Close()
+		if !op.Done() {
+			fmt.Fprintf(os.Stderr, "Waiting for the deletion of the certificate authority %q\n", caName)
+			if _, err := op.Wait(ctx); err != nil {
+				fmt.Fprintf(os.Stderr, "Error deleting certificate authority %q: %v\n", caName, err)
+				return err
+			}
+		}
+		fmt.Fprintf(os.Stderr, "Done deleting certificate authority %q\n", caName)
+		return nil
+	}, nil
+}
+
+// deleteCAPool deletes the Service Weaver certificate authority pool.
+// REQUIRES: the pool doesn't contain any certificate authorities.
+func deleteCAPool(ctx context.Context, config CloudConfig) (func() error, error) {
+	path := fmt.Sprintf("projects/%s/locations/%s/caPools/%s", config.Project, caLocation, caPoolName)
+
+	fmt.Fprintf(os.Stderr, "Fetching certificate authority pool %q... ", caPoolName)
+	client, err := privateca.NewCertificateAuthorityClient(ctx, config.ClientOptions()...)
+	if err != nil {
+		return nil, doErr(err)
+	}
+	_, err = client.GetCaPool(ctx, &privatecapb.GetCaPoolRequest{Name: path})
+	if isNotFound(err) {
+		client.Close()
+		fmt.Fprintf(os.Stderr, "Already deleted\n")
+		return nil, nil
+	} else if err != nil {
+		client.Close()
+		return nil, doErr(err)
+	}
+
+	fmt.Fprintf(os.Stderr, "Done. Starting async deletion... ")
+	op, err := client.DeleteCaPool(ctx, &privatecapb.DeleteCaPoolRequest{
+		Name: fmt.Sprintf("projects/%s/locations/%s/caPools/%s", config.Project, caLocation, caPoolName),
+	})
+	if err != nil {
+		client.Close()
+		return nil, doErr(err)
+	}
+	fmt.Fprintf(os.Stderr, "Done\n")
+	return func() error {
+		defer client.Close()
+		if !op.Done() {
+			fmt.Fprintf(os.Stderr, "Waiting for the deletion of the certificate authority pool %q\n", caPoolName)
+			if err := op.Wait(ctx); err != nil {
+				fmt.Fprintf(os.Stderr, "Error deleting certificate authority pool %q: %v\n", caPoolName, err)
+				return err
+			}
+		}
+		fmt.Fprintf(os.Stderr, "Done deleting certificate authority pool %q\n", caPoolName)
+		return nil
+	}, nil
+}
+
+// handleDNSZone handles the deletion of the given Service Weaver DNS zone
+// and its DNS records.
+func handleDNSZone(ctx context.Context, config CloudConfig, zone string) error {
+	z, err := getDNSZone(ctx, config, zone)
+	if err != nil {
+		return err
+	}
+	if z == nil { // already deleted
+		return nil
+	}
+
+	// Delete DNS record sets inside the DNS zone, as the zone must be empty
+	// before it can be deleted.
+	records, err := getDNSRecordSets(ctx, config, zone)
+	if err != nil {
+		return nil
+	}
+	if len(records) == 0 { // already deleted
```

Comment:
I'm not familiar with DNS zones, but is it possible to have a DNS zone without any records?